WordPress/gutenberg
# Add .wp-env.json support to wp-env

Adds support to `wp-env` for `.wp-env.json` files as described in https://github.com/WordPress/gutenberg/issues/17724#issuecomment-564952966. 

![wp-env](https://user-images.githubusercontent.com/612155/73631845-40954680-46ae-11ea-8857-a3111148d116.gif)

### Specification

`.wp-env.json` supports three fields:

| Field | Type | Default | Description |
| -- | -- | -- | -- |
| `"core"` | `string` or `null` | `null` | The WordPress installation to use. If `null` is specified, `wp-env` will use the latest production release of WordPress. |
| `"plugins"` | `string[]` | `[]` | A list of plugins to install and activate in the environment. |
| `"themes"` | `string[]` | `[]` | A list of themes to install in the environment. The first theme in the list will be activated. |

Several types of strings can be passed into these fields:

| Type | Format | Example(s) |
| -- | -- | -- |
| Relative path | `.<path>` or `~<path>` | `"./a/directory"`, `"../a/directory"`, `"~/a/directory"` |
| Absolute path | `/<path>` or `<letter>:\<path>` | `"/a/directory"`, `"C:\\a\\directory"` |
| GitHub repository | `<owner>/<repo>[#<ref>]` | `"WordPress/WordPress"`, `"WordPress/gutenberg#master"` |

Remote sources will be downloaded into a temporary directory located in `~/.wp-env`.

### Examples

#### Latest production WordPress + current directory as a plugin

This is useful for plugin development.

```json
{
  "core": null,
  "plugins": [
    "."
  ]
}
```

#### Latest development WordPress + current directory as a plugin

This is useful for plugin development when upstream Core changes need to be tested.

```json
{
  "core": "WordPress/WordPress#master",
  "plugins": [
    "."
  ]
}
```

#### Local `wordpress-develop` + current directory as a plugin

This is useful for working on plugins and WordPress Core at the same time.

```json
{
  "core": "../wordpress-develop/build",
  "plugins": [
    "."
  ]
}
```

#### A complete testing environment

This is useful for integration testing: that is, testing how old versions of WordPress and different combinations of plugins and themes impact each other.

```json
{
  "core": "WordPress/WordPress#5.2.0",
  "plugins": [
    "WordPress/wp-lazy-loading",
    "WordPress/classic-editor",
  ],
  "themes": [
    "WordPress/theme-experiments"
  ]
}
```

### Testing

1. Check out this branch
2. `cd` to a directory that contains a WordPress installation, a plugin, a theme, or a `.wp-env.json` file
3. Run `/path/to/gutenberg/packages/env/bin/wp-env start` 

### Questions

- Is the `wp-env reset` command useful? It has some problems, namely that it requires `wp-env start` to have been run at least once. Since we’re bumping the major version here it might be a good chance to replace `wp-env-reset` with e.g. `wp-env destroy` which destroys all containers and volumes so that one can run `wp-env start` from a clean slate. This would be useful for troubleshooting. 
- What is the `tests-wordpress-phpunit` container used for? I could not find where this is used and so have left it out for now.
